File: go/pingpong/packet.go
package main

import (
	"encoding/hex"
	"fmt"
	"strconv"
	"strings"
	"time"
)

// BitVector
type BitVector struct {
	Path []byte
}

func NewBitVector(hexStr string) BitVector {
	if len(hexStr)%2 != 0 {
		hexStr = "0" + hexStr
	}
	bytes, _ := hex.DecodeString(hexStr)
	return BitVector{Path: bytes}
}

func (bv *BitVector) String() string {
	return strings.ToUpper(hex.EncodeToString(bv.Path))
}

// Channel
type Channel struct {
	Path   BitVector
	Layer  int
	Signal int
}

func NewChannelFromString(s string) Channel {
	parts := strings.Split(s, "|")
	pathStr := ""
	if len(parts) > 0 {
		pathStr = parts[0]
	}
	layer := 0
	if len(parts) > 1 {
		layer, _ = strconv.Atoi(parts[1])
	}
	signal := 0
	if len(parts) > 2 {
		signal, _ = strconv.Atoi(parts[2])
	}
	return Channel{Path: NewBitVector(pathStr), Layer: layer, Signal: signal}
}

func (c Channel) String() string {
	return fmt.Sprintf("%s|%d|%d", c.Path.String(), c.Layer, c.Signal)
}

// Status
type Status struct {
	Age    int
	Health string
}

func NewStatusFromString(s string) Status {
	parts := strings.Split(s, "|")
	age := 0
	if len(parts) > 0 {
		age, _ = strconv.Atoi(parts[0])
	}
	health := "0000"
	if len(parts) > 1 {
		health = parts[1]
	}
	return Status{Age: age, Health: health}
}

func (s Status) String() string {
	return fmt.Sprintf("%d|%s", s.Age, s.Health)
}

// Peering
type Peering struct {
	Timestamp int64
	FromPeer  string
	ToPeer    string
	Subject   string
}

func NewPeeringFromString(s string) Peering {
	parts := strings.Split(s, " ")
	ts, _ := strconv.ParseInt(parts[0], 10, 64)
	return Peering{Timestamp: ts, FromPeer: parts[1], ToPeer: parts[2], Subject: parts[3]}
}

func (p Peering) String() string {
	return fmt.Sprintf("%d %s %s %s", p.Timestamp, p.FromPeer, p.ToPeer, p.Subject)
}

// Transformer
type Transformer struct {
	Pointer   int
	Operators []byte
	Data      string
}

func NewTransformerFromStrings(lines []string) Transformer {
	meta := lines[0]
	parts := strings.Split(meta, "|")
	ptr, _ := strconv.Atoi(parts[0])

	var ops []byte
	if len(parts) > 1 && parts[1] != "" {
		for _, opStr := range strings.Split(parts[1], ",") {
			if opStr != "" {
				op, _ := strconv.Atoi(opStr)
				ops = append(ops, byte(op))
			}
		}
	}
	data := strings.Join(lines[1:], "\n")
	return Transformer{Pointer: ptr, Operators: ops, Data: data}
}

func (t Transformer) String() string {
	var opsStrs []string
	for _, op := range t.Operators {
		opsStrs = append(opsStrs, strconv.Itoa(int(op)))
	}
	return fmt.Sprintf("%d|%s\n%s", t.Pointer, strings.Join(opsStrs, ","), t.Data)
}

// Packet
type Packet struct {
	Channel     Channel
	Status      Status
	Peering     Peering
	Transformer Transformer
}

func NewPacketFromString(s string) Packet {
	lines := strings.Split(s, "\n")
	return Packet{
		Channel:     NewChannelFromString(lines[0]),
		Status:      NewStatusFromString(lines[1]),
		Peering:     NewPeeringFromString(lines[2]),
		Transformer: NewTransformerFromStrings(lines[3:]),
	}
}

func (p Packet) String() string {
	return fmt.Sprintf("%s\n%s\n%s\n%s", p.Channel, p.Status, p.Peering, p.Transformer)
}

// Helper Packeter
func CreateSignal(from, to, subject, data string) string {
	p := Packet{
		Channel: Channel{Path: NewBitVector(""), Layer: 0, Signal: 1},
		Status:  Status{Age: 0, Health: "0000"},
		Peering: Peering{
			Timestamp: time.Now().UnixMilli(),
			FromPeer:  from,
			ToPeer:    to,
			Subject:   subject,
		},
		Transformer: Transformer{Pointer: 0, Operators: []byte{0}, Data: data},
	}
	return p.String()
}

File: java/pingpng/src/main/java/Packet.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

public class Packet {

    public static class BitVector {
        public byte[] path;

        public BitVector(String hexStr) {
            if (hexStr == null || hexStr.isEmpty()) {
                this.path = new byte[0];
                return;
            }
            if (hexStr.length() % 2 != 0) {
                hexStr = "0" + hexStr;
            }
            int len = hexStr.length();
            byte[] data = new byte[len / 2];
            for (int i = 0; i < len; i += 2) {
                data[i / 2] = (byte) ((Character.digit(hexStr.charAt(i), 16) << 4)
                                     + Character.digit(hexStr.charAt(i+1), 16));
            }
            this.path = data;
        }

        @Override
        public String toString() {
            StringBuilder sb = new StringBuilder();
            for (byte b : path) {
                sb.append(String.format("%02X", b));
            }
            return sb.toString();
        }
    }

    public static class Channel {
        public BitVector path;
        public long layer;
        public long signal;

        public Channel(String pathStr, long layer, long signal) {
            this.path = new BitVector(pathStr);
            this.layer = layer;
            this.signal = signal;
        }

        public static Channel fromString(String s) {
            String[] parts = (s != null ? s : "").split("\\|");
            String pathStr = parts.length > 0 ? parts[0] : "";
            long layer = parts.length > 1 && !parts[1].isEmpty() ? Long.parseLong(parts[1]) : 0;
            long signal = parts.length > 2 && !parts[2].isEmpty() ? Long.parseLong(parts[2]) : 0;
            return new Channel(pathStr, layer, signal);
        }

        @Override
        public String toString() {
            return path.toString() + "|" + layer + "|" + signal;
        }
    }

    public static class Status {
        public long age;
        public String health;

        public Status(long age, String health) {
            this.age = age;
            this.health = health;
        }

        public static Status fromString(String s) {
            String[] parts = (s != null ? s : "").split("\\|");
            long age = parts.length > 0 && !parts[0].isEmpty() ? Long.parseLong(parts[0]) : 0;
            String health = parts.length > 1 ? parts[1] : "0000";
            return new Status(age, health);
        }

        @Override
        public String toString() {
            return age + "|" + health;
        }
    }

    public static class Peering {
        public long timestamp;
        public String fromPeer;
        public String toPeer;
        public String subject;

        public Peering(long timestamp, String fromPeer, String toPeer, String subject) {
            this.timestamp = timestamp;
            this.fromPeer = fromPeer;
            this.toPeer = toPeer;
            this.subject = subject;
        }

        public static Peering fromString(String s) {
            String[] parts = (s != null ? s : "").split(" ");
            long ts = parts.length > 0 && !parts[0].isEmpty() ? Long.parseLong(parts[0]) : 0;
            String from = parts.length > 1 ? parts[1] : "";
            String to = parts.length > 2 ? parts[2] : "";
            String subj = parts.length > 3 ? parts[3] : "";
            return new Peering(ts, from, to, subj);
        }

        @Override
        public String toString() {
            return timestamp + " " + fromPeer + " " + toPeer + " " + subject;
        }
    }

    public static class Transformer {
        public long pointer;
        public byte[] operators;
        public String data;

        public Transformer(long pointer, byte[] operators, String data) {
            this.pointer = pointer;
            this.operators = operators;
            this.data = data;
        }

        public static Transformer fromStrings(List<String> lines) {
            String metaLine = lines.size() > 0 ? lines.get(0) : "0|0";
            String[] parts = metaLine.split("\\|");
            long ptr = parts.length > 0 && !parts[0].isEmpty() ? Long.parseLong(parts[0]) : 0;
            
            List<Byte> ops = new ArrayList<>();
            if (parts.length > 1 && !parts[1].isEmpty()) {
                for (String opStr : parts[1].split(",")) {
                    if (!opStr.trim().isEmpty()) {
                        ops.add((byte) Integer.parseInt(opStr.trim()));
                    }
                }
            } else {
                ops.add((byte) 0);
            }

            byte[] opsArr = new byte[ops.size()];
            for (int i = 0; i < ops.size(); i++) opsArr[i] = ops.get(i);

            String dataStr = lines.size() > 1 ? String.join("\n", lines.subList(1, lines.size())) : "";
            return new Transformer(ptr, opsArr, dataStr);
        }

        @Override
        public String toString() {
            StringBuilder sb = new StringBuilder();
            for (int i = 0; i < operators.length; i++) {
                sb.append(operators[i]);
                if (i < operators.length - 1) sb.append(",");
            }
            return pointer + "|" + sb.toString() + "\n" + data;
        }
    }

    public Channel channel;
    public Status status;
    public Peering peering;
    public Transformer transformer;

    public Packet(Channel channel, Status status, Peering peering, Transformer transformer) {
        this.channel = channel;
        this.status = status;
        this.peering = peering;
        this.transformer = transformer;
    }

    public static Packet fromString(String s) {
        String[] lines = s.split("\r?\n");
        Channel ch = Channel.fromString(lines.length > 0 ? lines[0] : "");
        Status st = Status.fromString(lines.length > 1 ? lines[1] : "");
        Peering pr = Peering.fromString(lines.length > 2 ? lines[2] : "");
        
        List<String> tLines = new ArrayList<>();
        for (int i = 3; i < lines.length; i++) {
            tLines.add(lines[i]);
        }
        Transformer tr = Transformer.fromStrings(tLines);

        return new Packet(ch, st, pr, tr);
    }

    @Override
    public String toString() {
        return channel.toString() + "\n" +
               status.toString() + "\n" +
               peering.toString() + "\n" +
               transformer.toString();
    }

    public static String createSignal(String from, String to, String subject, String data) {
        Packet p = new Packet(
            new Channel("", 0, 1),
            new Status(0, "0000"),
            new Peering(System.currentTimeMillis(), from, to, subject),
            new Transformer(0, new byte[]{0}, data)
        );
        return p.toString();
    }
}